### PR TITLE
Fetch Region and CLUSTER_ID information from cni-metrics-helper env

### PR DESCRIPTION
**Pull request by cgchinmay**
**What type of PR is this?**
feature

**Which issue does this PR fix**:
Allows Customers to use cni-metrics-helper by completely blocking access to IMDS

**What does this PR do / Why do we need it**:
Removes hard dependency of cni-metrics-helper on IMDS

**If an issue # is not available please add repro steps and logs from IPAMD/CNI showing the issue**:
#1287 

**Testing done on this change**:
**Automation added to e2e**:
Added Unit test
Manually tested on sample cluster

Tested by disabling [pod-imds ](https://docs.aws.amazon.com/eks/latest/userguide/best-practices-security.html#restrict-ec2-credential-access)
```
"MetadataOptions": {
                        "State": "applied",
                        "HttpTokens": "required",
                        "HttpPutResponseHopLimit": 1,
                        "HttpEndpoint": "enabled"
                    },
```
**Will this break upgrades or downgrades. Has updating a running cluster been tested?**:
No

**Does this change require updates to the CNI daemonset config files to work?**:

Yes, if IMDS access needs to be blocked

**Does this PR introduce any user-facing change?**:
Yes

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cmd/cni-metrics-helper/metrics/metrics.go [hunk 1]
```diff
@@ -317,7 +317,9 @@ func produceCloudWatchMetrics(t metricsTarget, families map[string]*dto.MetricFa
 	for key, family := range families {
 		convertMetrics := convertDef[key]
 		mType := family.GetType()
+		log.Infof("mType: %v", mType)
```
Comment (achevuru) [inline cmd/cni-metrics-helper/metrics/metrics.go]: Can we replace `mType` with something that explains itself?

File: cmd/cni-metrics-helper/metrics/metrics.go [hunk 2]
```diff
@@ -326,6 +328,7 @@ func produceCloudWatchMetrics(t metricsTarget, families map[string]*dto.MetricFa
 						Unit:       aws.String(cloudwatch.StandardUnitCount),
 						Value:      aws.Float64(action.data.curSingleDataPoint),
 					}
+					log.Infof("Publishing...")
```
Comment (achevuru) [inline cmd/cni-metrics-helper/metrics/metrics.go]: Do we need this additional log?

File: pkg/publisher/publisher.go [hunk 3]
```diff
@@ -85,21 +85,31 @@ type cloudWatchPublisher struct {
...
+			clusterID = getClusterID(ec2Client)
+		}
+
+		// Get ec2metadata client
+		ec2MetadataClient := ec2metadatawrapper.New(sess)
 
-	region, err := ec2MetadataClient.Region()
-	if err != nil {
-		return nil, errors.Wrap(err, "publisher: Unable to obtain region")
+		region, err = ec2MetadataClient.Region()
```
Comment (achevuru) [inline pkg/publisher/publisher.go:109]: Why do we want to do this if we already have `region` available?

File: cmd/cni-metrics-helper/metrics/metrics.go (hunk 2 above)
Comment (cgchinmay) [reply cmd/cni-metrics-helper/metrics/metrics.go]: I removed it, by mistake it got pushed

File: pkg/publisher/publisher.go (hunk 3 above)
Comment (cgchinmay) [reply pkg/publisher/publisher.go:109]: It might not be available if we are not using IRSA. I will update this line to fetch only if we dont have region

Comment (achevuru) [reply pkg/publisher/publisher.go:109]: Yes, similar to the `ClusterID` check you have above.

**Review by jayanthvn — CHANGES_REQUESTED** [review]
Can you also add a section here - https://github.com/aws/amazon-vpc-cni-k8s/tree/master/cmd/cni-metrics-helper. We can give steps on how to enable IRSA for cni-metrics helper and include this - 

```
Case 1: Cx not using IRSA, we need to get region and clusterID using IMDS (existing approach)
 	// Case 2: Cx using IRSA but not specified clusterID, we can still get this info if IMDS is not blocked
 	// Case 3: Cx blocked IMDS access and not using IRSA (which means region == "") OR
 	// not specified clusterID then its a Cx error
```

**Comment by jayanthvn** [comment]
Also as part of this, lets update https://docs.aws.amazon.com/eks/latest/userguide/cni-metrics-helper.html for IRSA support.

File: config/master/manifests.jsonnet [hunk 4]
```diff
@@ -297,40 +297,11 @@ local metricsHelper = {
       {
         apiGroups: [""],
         resources: [
-          "nodes",
```
Comment (jayanthvn) [inline config/master/manifests.jsonnet:300]: This needs to be updated too - https://github.com/aws/amazon-vpc-cni-k8s/blob/master/charts/cni-metrics-helper/templates/clusterrole.yaml

File: pkg/publisher/publisher.go [hunk 5]
```diff
@@ -85,21 +85,33 @@ type cloudWatchPublisher struct {
...
-	region, err := ec2MetadataClient.Region()
-	if err != nil {
-		return nil, errors.Wrap(err, "publisher: Unable to obtain region")
+		// If Customers have explicitly specified clusterID then skip auto generating it
+		if clusterID == "" {
+			clusterID = getClusterID(ec2Client)
+		}
+
+		// Get ec2metadata client
+		ec2MetadataClient := ec2metadatawrapper.New(sess)
```
Comment (achevuru) [inline pkg/publisher/publisher.go:107]: Why do this outside of the below `if` loop? We don't need to do this if `region` is already available isn't it

File: pkg/publisher/publisher.go [hunk 6]
```diff
@@ -93,24 +93,25 @@ func New(ctx context.Context, region string, clusterID string) (Publisher, error
 	// Case 3: Cx blocked IMDS access and not using IRSA (which means region == "") OR
 	// not specified clusterID then its a Cx error
 	if region == "" || clusterID == "" {
```
Comment (achevuru) [inline pkg/publisher/publisher.go:95]: Minor nit around code structure. 

Code is structured as below right now.
`
if region == "" || clusterID == ""
{
   if clusterID == ""{
   }
   
   if region == "" {
   }
}
`
Why do we need the upper level `if region == "" || clusterID == ""`? Instead below will suffice right? Prior to this diff update, we had few additional things we were doing outside `if region == ""` and `if clusterID == ""`blocks but with the below changes, everything is confined to both the individual `if` blocks..

`
 if clusterID == ""{
   }
   
   if region == "" {
   }
`
From what I can see, looks like something we don't really need the upper check.

Comment (cgchinmay) [reply pkg/publisher/publisher.go]: Yes, we dont need it. Have Updated the logic

**Comment by groodt** [comment]
Nice! Any ETA on `v1.10.2` being GA?

**Comment by cgchinmay** [comment]
> Nice! Any ETA on `v1.10.2` being GA?

Will soon have one, but will be around mid February. If you want to use this change then we do have private image release. Change the image tag to v1.10.2-rc1 and apply this manifest.

**Comment by cgchinmay** [comment]
> > Change the image tag to v1.10.2-rc1 and apply this manifest.
> 
> I've tried upgrading `amazon-vpc-cni-k8s` to v1.10.2-rc1 and unfortunately aws-node fails to start.
> 
> ```
> {"level":"info","ts":"2022-01-10T11:17:44.242Z","caller":"entrypoint.sh","msg":"Validating env variables ..."}
> {"level":"info","ts":"2022-01-10T11:17:44.339Z","caller":"entrypoint.sh","msg":"Install CNI binaries.."}
> {"level":"info","ts":"2022-01-10T11:17:44.644Z","caller":"entrypoint.sh","msg":"Starting IPAM daemon in the background ... "}
> {"level":"info","ts":"2022-01-10T11:17:44.739Z","caller":"entrypoint.sh","msg":"Checking for IPAM connectivity ... "}
> {"level":"info","ts":"2022-01-10T11:17:47.041Z","caller":"entrypoint.sh","msg":"Retrying waiting for IPAM-D"}
> panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x1 addr=0x39 pc=0x55ed00eac7c8]
> 
> goroutine 552 [running]:
> github.com/aws/amazon-vpc-cni-k8s/pkg/ipamd.(*IPAMContext).StartNodeIPPoolManager(0x0)
> 	/go/src/github.com/aws/amazon-vpc-cni-k8s/pkg/ipamd/ipamd.go:633 +0x28
> created by main._main
> 	/go/src/github.com/aws/amazon-vpc-cni-k8s/cmd/aws-k8s-agent/main.go:64 +0x32c
> {"level":"info","ts":"2022-01-10T11:17:49.540Z","caller":"entrypoint.sh","msg":"Retrying waiting for IPAM-D"}
> ```

Add these env vars to your aws-node ds
- name: ENABLE_IPv4
 value: "true"
 - name: ENABLE_IPv6
 value: "false"

Or instead apply the manifest for [v1.10.1](https://raw.githubusercontent.com/aws/amazon-vpc-cni-k8s/v1.10.1/config/master/aws-k8s-cni.yaml) first and then change the image tag.

**Comment by groodt** [comment]
> Or instead apply the manifest for [v1.10.1](https://raw.githubusercontent.com/aws/amazon-vpc-cni-k8s/v1.10.1/config/master/aws-k8s-cni.yaml) first and then change the image tag.

Thanks! It worked.

**Comment by jayanthvn** [comment]
@cgchinmay - Can you update your branch? We can finish the review this week.

File: cmd/cni-metrics-helper/README.md [hunk 7]
```diff
@@ -13,6 +13,94 @@ The following diagram shows how `cni-metrics-helper` works in a cluster:
...
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: cni-metrics-helper
+subjects:
+  - kind: ServiceAccount
+    name: <IRSA name>
+    namespace: kube-system
+```
+
+3. Specify this IRSA in the cni-metrics-helper deployment spec alongwith CLUSTER_ID as the metric dimension
```
Comment (jayanthvn) [inline cmd/cni-metrics-helper/README.md:68]: Nit: Can you expand on what can be set for AWS_CLUSTER_ID? Maybe you can add a section in readme like this https://github.com/aws/amazon-vpc-cni-k8s#cluster_name

**Review by jayanthvn — APPROVED** [review]
Just one minor nit, rest lgtm.

File: cmd/cni-metrics-helper/README.md (hunk 7 above)
Comment (cgchinmay) [reply cmd/cni-metrics-helper/README.md]: Updated. have a look :)

Comment (jayanthvn) [reply cmd/cni-metrics-helper/README.md]: Looks good, thanks.

https://github.com/cgchinmay/amazon-vpc-cni-k8s/tree/fix_imds_access_cw/cmd/cni-metrics-helper

**Pull request merged**